lib/scholar/metrics/regression.ex

```diff
@@ -642,4 +642,253 @@ defmodule Scholar.Metrics.Regression do
+      #Nx.Tensor<
+        f32
+        0.7727271914482117
+      >
+      iex> Scholar.Metrics.Regression.d2_pinball_score(y_true, y_true, alpha: 0.1)
+      #Nx.Tensor<
+        f32
+        1.0
+      >
+  """
+
+  deftransform d2_pinball_score(y_true, y_pred, opts \\ []) do
+    if Nx.size(y_pred) < 2 do
+      Nx.Constants.nan()
+    else
+      d2_pinball_score_n(y_true, y_pred, NimbleOptions.validate!(opts, @d2_pinball_score_schema))
+    end
+  end
+
+  defnp d2_pinball_score_n(y_true, y_pred, opts \\ []) do
+    alpha = opts[:alpha]
+    shape = Nx.shape(y_true)
+
+    m =
+      if Nx.rank(shape) == 1 do
+        shape |> elem(0)
+      else
+        shape |> elem(1)
+      end
+
+    numerator = mean_pinball_loss(y_true, y_pred, alpha: alpha, multioutput: :raw_values)
+
+    y_quantile = Nx.broadcast(quantile(y_true, alpha), shape)
+    denominator = mean_pinball_loss(y_true, y_quantile, alpha: alpha, multioutput: :raw_values)
+
+    nonzero_numerator = numerator != 0
+    nonzero_denominator = denominator != 0
+
+    valid_score = nonzero_numerator and nonzero_denominator
+    invalid_score = nonzero_numerator and not nonzero_denominator
+
+    output_scores = Nx.broadcast(1, {m})
+
+    output_scores =
+      Nx.select(valid_score, Nx.subtract(1, Nx.divide(numerator, denominator)), output_scores)
+
+    output_scores = Nx.select(invalid_score, 0.0, output_scores)
+
+    case opts[:multioutput] do
+      :uniform_average -> Nx.mean(output_scores)
+      :raw_values -> output_scores
+    end
+  end
+
+  defn quantile(tensor, q) do
+    rank = Nx.rank(tensor)
+
+    if rank == 1 do
+      sorted_tensor = Nx.sort(tensor)
+      n = Nx.size(sorted_tensor)
+      float_index = q * (n - 1)
+      index = Nx.floor(float_index) |> Nx.as_type(:s64)
+      next_index = Nx.min(index + 1, n - 1)
+
+      value_at_index = Nx.take(sorted_tensor, index)
+      value_at_next_index = Nx.take(sorted_tensor, next_index)
+
+      weight = float_index - Nx.floor(float_index)
+
+      quantile =
+        Nx.add(
+          Nx.multiply(value_at_index, 1 - weight),
+          Nx.multiply(value_at_next_index, weight)
+        )
+
+      Nx.squeeze(quantile)
+    else
+      sorted_tensor = Nx.sort(tensor, axis: 0)
+      {n, m} = Nx.shape(sorted_tensor)
+
+      float_indices = q * (n - 1)
+      indices = Nx.floor(float_indices) |> Nx.as_type(:s64)
+      next_indices = Nx.min(indices + 1, n - 1)
+
+      indices_tensor = Nx.broadcast(indices, {n, m})
+      next_indices_tensor = Nx.broadcast(next_indices, {n, m})
+
+      values_at_indices = Nx.take_along_axis(sorted_tensor, indices_tensor, axis: 0)
+      values_at_next_indices = Nx.take_along_axis(sorted_tensor, next_indices_tensor, axis: 0)
+
+      weights = float_indices - Nx.floor(float_indices)
+
+      quantiles =
+        Nx.add(
+          Nx.multiply(values_at_indices, 1 - weights),
+          Nx.multiply(values_at_next_indices, weights)
+        )
+
+      Nx.squeeze(quantiles)
+    end
```

Comment:
This fragment might be consolidated, because there is a repeating code. The first 5 lines and last 3 are identical in both branches of `if`.